pkg/api/validation/validation_test.go

```diff
@@ -1866,6 +1805,160 @@ func TestValidatePod(t *testing.T) {
 				DNSPolicy:     api.DNSClusterFirst,
 			},
 		},
+		{ // Serialized pod affinity in affinity requirements in annotations.
+			ObjectMeta: api.ObjectMeta{
+				Name:      "123",
+				Namespace: "ns",
+				// TODO: Uncomment and move this block into Annotations map once
+				// RequiredDuringSchedulingRequiredDuringExecution is implemented
+				//"requiredDuringSchedulingRequiredDuringExecution": [
+				//	{
+				//		"labelSelector":
+				//			{
+				//				"matchExpressions": [{
+				//					"key": "key2",
+				//					"operator": "In",
+				//					"values": ["value1", "value2"]
+				//				}]
+				//			},
+				//			"namespaces":[
+				//			{
+				//				"metadata":{
+				//					"name": "ns",
+				//					"labels":{
+				//						"key1": "value1",
+				//						"key2": "value2"
+				//					}
+				//				}
+				//			}],
+				//			"topologyKey": "zone"
+				//		}
+				//	]
+				Annotations: map[string]string{
+					api.AffinityAnnotationKey: `
+					{"podAffinity": {
+						"requiredDuringSchedulingIgnoredDuringExecution": [{
```

Comment:
Just to be clear, affinity being expressed as an annotation is only temporary. It will become a field of PodSpec once we are comfortable with the feature and can move it out of alpha (presumably no later than the 1.4 release). It's certainly true that making it an annotation, even temporarily, is not ideal, but because Pod is already v1, the only way to do an alpha version of affinity is to make it a completely new API resource, or do what we did here (make it an annotation). Making it a completely new API resource has a lot of drawbacks and few benefits.

As for the format of the struct itself (ignoring annotation vs. field) -- I don't think it qualifies as a DSL. The "hard" requirements are really just LabelSelectors (which have been in the system for some time now) paired with the name of a label key ("TopologyKey"), and the "soft" requirements are just "hard" requirements plus a weight. And the hard requirements can only be combined by simple AND (i.e. listing more than one rule implies AND). So it's actually not very expressive IMO.